### ISSUE TITLE
Hello, any motivation to resume on this ?

### ISSUE THREAD

**Issue by Olfried2** [eltariel/njak#1]
I'm interested, but I'm a pyton noob.
I could help with manuals, graphics and other linux related stuff from the view of an internediate user.

**Comment by eltariel** [comment]
Yes, definitely still motivation, I've been working on it on and off. I've kind of changed track with it recently - the button and light control is still there but I've temporarily dropped the USB HID in favour of an MQTT interface because I needed a physical control for my smart lights and I don't have another keybow to work with...

I've got a few things I want to do with the code, in no particular order:
- bring back USB HID
- move USB HID init into the main app
- add Bluetooth HID
- document all of this craziness
- fix the project layout, add testing and linting etc now that I have some more experience with python
- build a minimal Linux distro to run this on, similar to the official keybow firmware

I'm more than happy to have some help with this, or add other stuff, or... anything 😄

**Comment by Olfried2** [comment]
Hello Eltariel,
as starting point, I'd like to suggest "dietPie.com" which I use on all of my Rasberries. I think it's not necessary to
make an special OS for the keybow. As I know that one could start python on "startup" (even I didn't got it working)
It could be nice. Also if you've an raspberry with wlan, it should be possible to connect via ssh.
But for the cheap solution, it would be also acceptable to connect via serial connection.
My main topics to use the keybow are:
sending ascii text , sentences or phrases, e-mail adresses(the actual software from pimoroni doesn't support the sending of "@" and there's no documentation on how to change the keyboard-map (even that they have some other languages than english). 
It would be great, if there were layers [ but one could change the settings using the serial input, instead].
And of course changing the function to MIDI, on the fly, would be great. I'd like to use the keybow also to control my ardour DAW.
For me, the LED's would be an indicator which layer is chosen and which key has been pressed or locked.
....
"Bluetooth HID" not on the first page of my wishes
"Documentation", That's where I could come up! (also refreshes my noob nowledge of python)
"testing..." Also a part I can do

I can provide the german keyboard layout (scancodes)
...
Looking forward, keep healthy

**Comment by eltariel** [comment]
For the record, I started this whole project at least as much as an excuse to learn python as anything else. It kind of doesn't matter if you're a noob, we've got to start somewhere!

- I had planned to look into something like dietpi or picore and just drop my app on top of one of them. I already have a working install on a raspbian lite image, but it still takes a bit longer to boot than I would like.
- Sending text strings is pretty simple. I had it mostly working at one point, but I let that bit of code die because I never used it... the most complex part of it is managing the modifier key states.
- Somewhere in there is super basic support for layers - designate one button as the layer select key, hold that and tap another key to switch. It effectively gives an 11x11 matrix of possible behaviours, didn't have that many ideas to assign to it...
- On-the-fly reprogramming (beyond layer changes, anyway) will need a lot of thought to get it in a useful state.
- I *also* had plans to use it as a MIDI controller for my DAW, but the priority on that dropped when I got a launchpad. Still wouldn't mind getting that going though.
- The USB setup script already creates interfaces for HID, RNDIS (networking), ACM (serial), and MIDI. I figure there's not really a lot of point using the serial connection when I can just expose a control interface over the network link, especially since I have the wifi as well.
- Bluetooth isn't a high priority, and it's certainly not trivial either. The nice bit though is that once you get the hardware in the right state it's the same HID as you'd use on USB.

At this point it's probably worth creating a new issue for each item on the wishlist... :P

**Comment by Olfried2** [comment]
Just ordered 2 Raspberry Pi Zero W, to have WLAN :)
Tomorrow, I'll make a sketch to visualize some Topics and their relations (freemind OK for You?)
